File: archive/flutter-app/test/core/utility/password_strength_analyzer_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:infrakit_studio/core/utility/password_strength_analyzer.dart';

void main() {
  const analyzer = PasswordStrengthAnalyzer();

  test('rates an obviously weak short password as weak', () {
    final result = analyzer.execute(const PasswordStrengthInput(password: 'abc'));
    expect(result.rating, PasswordStrengthRating.weak);
  });

  test('rates an obviously strong long mixed-charset password highly', () {
    final result = analyzer.execute(
      const PasswordStrengthInput(password: r'aB3!xQ9$zP2#Lm7^Rt5&Wc8@'),
    );
    expect(
      result.rating,
      anyOf(PasswordStrengthRating.strong, PasswordStrengthRating.veryStrong),
    );
  });

  test('rating ordering: weak password scores lower than strong password', () {
    final weak = analyzer.execute(const PasswordStrengthInput(password: 'abc'));
    final strong = analyzer.execute(
      const PasswordStrengthInput(password: r'aB3!xQ9$zP2#Lm7^Rt5&Wc8@'),
    );
    expect(weak.entropyBits, lessThan(strong.entropyBits));
    expect(weak.rating.index, lessThan(strong.rating.index));
  });

  test('empty password is weak with zero entropy', () {
    final result = analyzer.execute(const PasswordStrengthInput(password: ''));
    expect(result.entropyBits, 0);
    expect(result.rating, PasswordStrengthRating.weak);
  });

  test('longer passwords of the same charset never score lower', () {
    final shorter = analyzer.execute(const PasswordStrengthInput(password: 'aaaa'));
    final longer = analyzer.execute(const PasswordStrengthInput(password: 'aaaaaaaa'));
    expect(longer.entropyBits, greaterThan(shorter.entropyBits));
  });

  test('adding character classes increases entropy for equal length', () {
    final lowerOnly = analyzer.execute(const PasswordStrengthInput(password: 'abcdefgh'));
    final mixed = analyzer.execute(const PasswordStrengthInput(password: 'aB3defg!'));
    expect(mixed.entropyBits, greaterThan(lowerOnly.entropyBits));
  });
}

File: app/src/core/utility/passwordStrengthAnalyzer.test.ts
import { describe, expect, it } from 'vitest'
import { PasswordStrengthAnalyzer, passwordStrengthRatingIndex } from './passwordStrengthAnalyzer'

describe('PasswordStrengthAnalyzer', () => {
  const analyzer = new PasswordStrengthAnalyzer()

  it('rates an obviously weak short password as weak', () => {
    const result = analyzer.execute({ password: 'abc' })
    expect(result.rating).toBe('weak')
  })

  it('rates an obviously strong long mixed-charset password highly', () => {
    const result = analyzer.execute({ password: 'aB3!xQ9$zP2#Lm7^Rt5&Wc8@' })
    expect(['strong', 'veryStrong']).toContain(result.rating)
  })

  it('rating ordering: weak password scores lower than strong password', () => {
    const weak = analyzer.execute({ password: 'abc' })
    const strong = analyzer.execute({ password: 'aB3!xQ9$zP2#Lm7^Rt5&Wc8@' })
    expect(weak.entropyBits).toBeLessThan(strong.entropyBits)
    expect(passwordStrengthRatingIndex(weak.rating)).toBeLessThan(passwordStrengthRatingIndex(strong.rating))
  })

  it('empty password is weak with zero entropy', () => {
    const result = analyzer.execute({ password: '' })
    expect(result.entropyBits).toBe(0)
    expect(result.rating).toBe('weak')
  })

  it('longer passwords of the same charset never score lower', () => {
    const shorter = analyzer.execute({ password: 'aaaa' })
    const longer = analyzer.execute({ password: 'aaaaaaaa' })
    expect(longer.entropyBits).toBeGreaterThan(shorter.entropyBits)
  })

  it('adding character classes increases entropy for equal length', () => {
    const lowerOnly = analyzer.execute({ password: 'abcdefgh' })
    const mixed = analyzer.execute({ password: 'aB3defg!' })
    expect(mixed.entropyBits).toBeGreaterThan(lowerOnly.entropyBits)
  })
})
